treeppl/rfc3-functions-vectorization.md

```diff
@@ -0,0 +1,80 @@
+# TreePPL RFC 3: Functions and Vectorization
+
+## Sample function declaration 
+
+	gtr = function(p: Probability[4], r: Rate[5]) => InstantaneousRateMatrix
+	{
+		// ... 
+	}
+
+## Function invocation
+
+Arguments to functions can be labeled (like R, Python...)
+Labelling is optional
+Arguments can have default values
+Reordering the named values should be possible
+mixing of keyword and positional arguments should be possible
+
+Example:
+
+	  Q = gtr(p = pi, r = rep(1, 6))
+	  Q = gtr(pi, rep(1, 6))
+	  Q = gtr(r = rep(1, 6), p = pi)
+	  Q = gtr(p = pi, rep(1, 6))
+	  
+## Variadic arguments
+
+	foo(...arr1) {
+	 // arr1 is an array with the arguments...
+	}
+	
+	foo(a = 1, b = 2)
+	foo(a = 1, b = 2, c = 3)
+		 
+## Vectorization
+
+	a = 0:10 // [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

Comment:
In some languages 10 would be included, in some not. I propose to be able to specify both, e. g.:

  * `0..10` = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
  * `0...10` = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]